backup/multicloud/client.go

```diff
@@ -0,0 +1,265 @@
+}
+
+type InitiateMultipartUploadResult struct {
+	Xmlns    string `xml:"xmlns,attr"`
+	Bucket   string `xml:"Bucket"`
+	Key      string `xml:"Key"`
+	UploadId string `xml:"UploadId"`
+}
+
+type UploadPartResult struct {
+	Xmlns      string `xml:"xmlns,attr"`
+	PartNumber int64  `xml:"PartNumber"`
+	ETag       string `xml:"ETag"`
+}
+
+type Part struct {
+	PartNumber int64  `xml:"PartNumber"`
+	ETag       string `xml:"ETag"`
+}
+
+type CompleteMultipartUpload struct {
+	Xmlns string `xml:"xmlns,attr"`
+	Part  []Part `xml:"Part"`
+}
+
+type CompleteMultipartUploadResult struct {
+	Xmlns    string `xml:"xmlns,attr"`
+	Location string `xml:"Location"`
+	Bucket   string `xml:"Bucket"`
+	Key      string `xml:"Key"`
+	ETag     string `xml:"ETag"`
+}
+
+func (c *Client) UploadObject(bucketName, objectKey string, data []byte) error {
+	p := path.Join("s3", bucketName, objectKey)
+	if err := c.request("PUT", p, data, nil, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) ListObject(bucketName string) (*ListObjectResponse, error) {
+	p := path.Join("s3", bucketName)
+	object := &ListObjectResponse{}
+	if err := c.request("GET", p, nil, object, nil); err != nil {
+		return nil, err
+	}
+	return object, nil
+}
+
+func (c *Client) RemoveObject(bucketName, objectKey string) error {
+	p := path.Join("s3", bucketName, objectKey)
+	if err := c.request("DELETE", p, nil, nil, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) InitMultiPartUpload(bucketName, objectKey string) (*InitiateMultipartUploadResult, error) {
+	p := path.Join("s3", bucketName, objectKey)
+	p += "?uploads"
+	out := &InitiateMultipartUploadResult{}
+	if err := c.request("PUT", p, nil, out, nil); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func (c *Client) UploadPart(bucketName, objectKey string, partNum int64, uploadId string, data []byte, size int64) (*UploadPartResult, error) {
+	log.Infof("upload part buf size:%d", len(data))
+	p := path.Join("s3", bucketName, objectKey)
+	p += fmt.Sprintf("?partNumber=%d&uploadId=%s", partNum, uploadId)
+	out := &UploadPartResult{}
+	reqSettingCB := func(req *httplib.BeegoHTTPRequest) error {
+		req.Header("Content-Length", strconv.FormatInt(size, 10))
+		req.SetTimeout(30*time.Minute, 30*time.Minute)
+		return nil
+	}
+	if err := c.request("PUT", p, data, out, reqSettingCB); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func (c *Client) CompleteMultipartUpload(
+	bucketName string,
+	objectKey string,
+	uploadId string,
+	input *CompleteMultipartUpload) (*CompleteMultipartUploadResult, error) {
+
+	p := path.Join("s3", bucketName, objectKey)
+	p += fmt.Sprintf("?uploadId=%s", uploadId)
+	out := &CompleteMultipartUploadResult{}
+	if err := c.request("PUT", p, input, nil, nil); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func (c *Client) AbortMultipartUpload(bucketName, objectKey string) error {
+	//p := path.Join("s3", "AbortMultipartUpload", bucketName, objectKey)
```

Comment:
Why are these commented out?  If this does not work yet, add a TODO to implement it later and delete unused code.

Reply:
done